src/unicode/capi.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * This file is part of the "libunicode" project
+ *   Copyright (c) 2021 Christian Parpart <christian@parpart.family>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if !defined(__cplusplus)
+extern "C" {
+#endif
+
+/// UTF-8 character or 8bit segment of an UTF-8 character.
+typedef char u8_char_t;
+
+/// UTF-32 codepoint.
+typedef uint32_t u32_char_t;
+
+/**
+ * Counts the number of grapheme clusters for given sequence of codepoints.
+ *
+ * Use this function to determine the number of
+ * user perceived characters (grapheme clusters).
+ *
+ * @param codepoints   pointer to the first codepoint.
+ * @param n            number of codepoints to count the grapheme clusters for.
+ *
+ * @return number of user perceived characters (grapheme clusters) counted
+ *         in [codepoints, codepoints+n).
+ */
+int u32_gc_count(u32_char_t const* codepoints, size_t n);
+int u8_gc_count(u8_char_t const* codepoints, size_t n);
+
+/**
+ * Computes the display width for given sequence of codepoints,
+ * respecting grapheme cluters, and modifiers.
+ *
+ * @param codepoints  pointer to first codepoint
+ * @param n           number of codepoints
+ * @param allowMod    booelan to indicate whether width modifications
+ *                    (e.g. via VS15/VS16) are allowed or not.
+ *
+ * Use this function to determine how many terminal grid cells a
+ * string of codepoints should occupy when being rendered.
+ */
+int u32_gc_width(u32_char_t const* codepoints, size_t n, int allowMod);
+
+/**
+ * UTF-8 version of @c u32_gc_width().
+ *
+ * @see u32_gc_width(u32_char_t const* codepoints, size_t n, int allowMod)
+ */
+int u8_gc_width(u8_char_t const* codepoints, size_t n, int allowMod);
+
+/**
+ * Tests if two consecutive codepoints do belong to the same grapheme cluster,
+ * i.e. are unbreakable and thus should not be broken up.
+ *
+ * @retval 1   both codepoints to belong to the same grapheme cluster.
+ * @retval 0   both codepoints do not belong to the same grapheme cluster.
+ */
+int u32_grapheme_unbreakable(u32_char_t a, u32_char_t b);
```

Comment:
I see, this could be used to get the cluster daisy-chaining for streams working, but to work reliable, one has either to do a backward search on the last chunk end, or get the last grapheme classification in the first place. How do you intend to deal with that?

My own solution in this regard was pretty ugly - it kinda carried a context object for streams around, that holds the needed information to stop/continue at arbitrary position. Well, I did not expose the breakable function to the outside, just a high level "cluster iterator", which could consume new data at any time.

Reply:
> I see, this could be used to get the cluster daisy-chaining for streams working, but to work reliable, one has either to do a backward search on the last chunk end, or get the last grapheme classification in the first place. How do you intend to deal with that?

See code sample above. IMHO that suffices.

> My own solution in this regard was pretty ugly - it kinda carried a context object for streams around, that holds the needed information to stop/continue at arbitrary position. Well, I did not expose the breakable function to the outside, just a high level "cluster iterator", which could consume new data at any time.

In Contour I am actually using the `unbreakable` function (C++ API). and if the last cursor positions right-most codepoint is "unbreakable" to the codepoint to be written, it then will be inserted into the last cursor position and the cursor itself is not altered. Otherwise the codepoint belongs to a new grapheme cluster and therefore will be placed as a (initially) single-codepoint sequence into the current cursor's cell and then the cursor incremented. -- What do you think, your implementation vs the one I just described?

Reply:
Well, thats the point I kinda stopped implementing it, thus this never got solved in my PR. Basically my thinking was the following:
- return whatever just arrived in terms of cluster segments, even the last prolly unfinished cluster (we do a utf8-->utf32 conversion prehand, which holds back half transmitted multibytes until next chunk, so we eliminated the "byte problem" here). This unfinished cell content might turn out problematic for crazy clusters, but will be repaired with "good continuation" by the next chunk.
- on next decoded chunk load back the "unfinished" context state (so far no cursor advance), and check in cluster algo, whether it continues correctly - add to active cell then advance normally, otherwise we got faulty continuation (always advance)

Reply:
NB: If I remember right the country flags were cumbersome in this regard. On byte level they are insanely long, thus likely to get cut by some transport buffer. If your implementation works correctly for any cut at any byte position, and can correctly add the codepoints into one cell, guess then it is fine. Same with the family emoji (didnt count it, would that qualify for >100 bytes in UTF8?).

Reply:
I do decode UTF-8 to 32 before hitting the VT parser so internally everything is UTF-32 Codepoints. Uhrengehäuse Populationen grapheme Clusters is what I do to and it works well. I think it should also be fine that way because your might encounter invalid Codepoints sequences auf you should still be able torender them efficiently.

Reply:
> NB: If I remember right the country flags were cumbersome in this regard. On byte level they are insanely long, thus likely to get split by some transport buffer. If your implementation works correctly for any cut at any byte position, and can correctly add the codepoints into one cell, guess then it is fine. Same with the family emoji (didnt count it, would that qualify for >100 bytes in UTF8?)

I can check that. I do not count in UTD-8. But the family emoji i am always using for tests is 7 UTF-32 Codepoints long.

If my implementation is not working somewhere, i will better go fix it then giving up. :-)

Reply:
7 codepoints? Ah well, then it is 28 bytes at max, what a pity :smile_cat: 

> If my implementation is not working somewhere, i will better go fix it rather than giving up. :-)

:+1: